misc/man/rrr_auth.1
.Dd 2020-09-16
.TH RRR_AUTH 1
.SH NAME
Read Route Record is a dynamic module-based message handling program
for collecting, buffering, processing and storing messages and measurements.
.PP
.B rrr_auth
authenticates users against a password file generated by  
.Xr rrr_passwd(1).
.SH SYNOPSIS
.B rrr_auth {PASSWORD_FILE} {USERNAME}
.Dl [-p|--permission[=]PERMISSION]
.Dl [-s|--stdin]
.Dl [-l|--loglevel-translation]
.Dl [-e|--environment-file[=]ENVIRONMENT FILE]
.Dl [-d|--debuglevel[=]DEBUG FLAGS]
.Dl [-D|--debuglevel-on-exit[=]DEBUG FLAGS]
.Dl [-h|--help]
.Dl [-v|--version]

.SH DESCRIPTION
This program authenticates a user according to given parameters.
.SH OPTIONS
.IP PASSWORD_FILE
The password file to use. Required parameter.
.IP USERNAME
The username to authenticate. Required parameter.
.IP -p|--permission[=]PERMISSION
A permission to require for the user. If not given, permissions are not checked for and only the password is validated.
.IP -s|--stdin
Read password from STDIN ending with EOF. If not given, terminal input with a prompt is performed.
.IP -d|--debuglevel[=]FLAGS
Debuglevel to use.
.IP -D|--debuglevel-on-exit[=]FLAGS
Debuglevel to use when program exits.
.IP -l|--loglevel-translation
Translate output loglevels to RFC5424 format. Internal loglevels will be translated so that error messages become ERROR
and other messages DEBUG (loglevels 3 and 7).
.IP -e|--environment-file
See
.Xr rrr(1)
for usage.

.SH RETURN VALUE
.B rrr_auth
returns 0 on successful authentication and 1 on failure. Nothing is printed upon successful authentication unless debuglevel 1 is set.
.PP
Read
.B SECURITY NOTES
below before enabling debugging.

.SH SECURITY NOTES
If debuglevel 1 is set (and possibly others), the program might expose whether a user exists or not.
This should be avoided if
.B rrr_auth
is used for for actual authentication and not just testing.

.SH DEBUG FLAGS
Refer to
.Xr rrr(1)
for possible debug flags. See also
.B RETURN VALUE
and
.B SECURITY NOTES.

.SH SEE ALSO
.Xr rrr(1),
.Xr rrr_passwd(1)


debian/changelog
rrr (1.13-1) buster; urgency=medium

  * Fix for ip module array byte by byte skipping which caused bug trap
  * Fixes for mmap channel shutdown hang when fork had been killed
  * In cmodule framwork, use ping-pong system to check if fork is alive
  * Fix in file module for keyboard events getting lost on slow HW
  * Fix for long sleeps in file module causing thread restart by WD
  * Fix for journald output not working on 32-bit
  * Misc. log message improvements

 -- Atle Solbakken <atle@goliathdns.no>  Mon, 21 Sep 2020 22:05:59 +0200

rrr (1.13-0) buster; urgency=medium

  * Implement native SystemD journal logging
  * Fix crash in high-load logging situtations in statistics engine
  * Misc. bugfixes and cleanups

 -- Atle Solbakken <atle@goliathdns.no>  Tue, 15 Sep 2020 18:31:55 +0200

rrr (1.12-2) buster; urgency=medium

  * Added RPM packages
  * Improvements in log messages / silence some messages
  * New parameters in file module file_max_messages_per_file and
    file_max_read_step_size
  * Fix in signal framework causing cmodule forks only to be killed
    by SIGKILL resulting in ungraceful shutdown
  * Fix for forks not getting initialized when there are lots (lots) of
    debug messages being printed
  * Minor fixes

 -- Atle Solbakken <atle@goliathdns.no>  Sat, 12 Sep 2020 16:42:46 +0200

rrr (1.12-1) buster; urgency=medium

  * Fix misc. build warnings

 -- Atle Solbakken <atle@goliathdns.no>  Fri, 04 Sep 2020 10:52:36 +0200

rrr (1.12-0) buster; urgency=medium

  * Implement LibreSSL as an alternative to OpenSSL

 -- Atle Solbakken <atle@goliathdns.no>  Thu, 03 Sep 2020 08:45:18 +0200

rrr (1.11-4) buster; urgency=medium
  
  * Fix bug with smart timeout in IP module

 -- Atle Solbakken <atle@goliathdns.no>  Thu, 03 Sep 2020 08:25:43 +0200

rrr (1.11-3) buster; urgency=medium

  * New configuration parameters in HTTP-server to allow push connections

 -- Atle Solbakken <atle@goliathdns.no>  Fri, 28 Aug 2020 15:08:45 +0200

rrr (1.11-2) buster; urgency=medium

  * Fix bug causing crash on arm64 when debuglevel 1 was active

 -- Atle Solbakken <atle@goliathdns.no>  Wed, 26 Aug 2020 23:56:57 +0200

rrr (1.11-1) buster; urgency=medium

  * New file module to read from devices, files, sockets etc.
  * Improvements in ip and socket framework
  * Misc. bugfixes

 -- Atle Solbakken <atle@goliathdns.no>  Wed, 26 Aug 2020 13:30:45 +0200

rrr (1.10-3) buster; urgency=medium

  * Fixes in IP module
    * Add TCP graylist timeout configuration parameter
    * Add configurable TTL check like buffer module
    * Add configurable smart timeout to update send of all messages to a single destination
    * Improve preserve order algorithm, send order is now guaranteed
    * Fix noisy debug messages
    * Fix that returned RRR messages was not endian converted back to host
    * Fix behaviour inconsistencies when using default target vs dynamic target
  * Improve fork shutdown algorithm
  * Misc fixes and cleanups

 -- Atle Solbakken <atle@goliathdns.no>  Sat, 22 Aug 2020 20:39:53 +0200

rrr (1.10-2) buster; urgency=medium

  * Fix MQTT array data parsing from PUBLISH which did not work

 -- Atle Solbakken <atle@goliathdns.no>  Mon, 17 Aug 2020 18:45:31 +0200

rrr (1.10-1) buster; urgency=medium

  * Fixes in array tree framework and IP module

 -- Atle Solbakken <atle@goliathdns.no>  Mon, 17 Aug 2020 00:44:07 +0200

rrr (1.10-0) buster; urgency=medium

  * New Array Tree framework for conditional array parsing (branching)
  * Implement HTTP raw modes in httpclient and httpserver
  * Misc smaller fixes and internal cleanups

 -- Atle Solbakken <atle@goliathdns.no>  Sun, 16 Aug 2020 13:37:24 +0200

rrr (1.9-2) buster; urgency=medium

  * Fixes in configuration file parsing, allow spaces after parameters
  * Fix proper connect retry vs restart in MQTT client
  * Other small fixes in MQTT client 

 -- Atle Solbakken <atle@goliathdns.no>  Thu, 06 Aug 2020 23:43:34 +0200

rrr (1.9-1) buster; urgency=medium

  * Merge security fix from 1.7-4
  * Fix performance problems and bugs in journal module
  * Fix performance problems in MQTT broker

 -- Atle Solbakken <atle@goliathdns.no>  Thu, 06 Aug 2020 11:37:29 +0200

rrr (1.9-0) buster; urgency=medium

  * Removed duplicator module, functionality is now in buffer module
  * Create array manipulation helper functions in Perl5
  * Remove old style array handling in Perl5
  * Add TTL check for messages in buffer module

 -- Atle Solbakken <atle@goliathdns.no>  Wed, 05 Aug 2020 12:35:40 +0200

rrr (1.8-1) buster; urgency=medium

  * Fix bug causing multiple configuration parameters not to work

 -- Atle Solbakken <atle@goliathdns.no>  Mon, 03 Aug 2020 17:44:39 +0200

rrr (1.8-0) buster; urgency=medium

  * Support will messages and retain in MQTT broker.
  * Misc. compliance fixes in MQTT broker.
  * Misc. fixes in connection handling in MQTT commonn
  * New configuration parameters in IP module to put timed out messages
    to output queue
  * Smaller packaging fixes (Debian)

 -- Atle Solbakken <atle@goliathdns.no>  Mon, 03 Aug 2020 14:56:38 +0200

rrr (1.7-4) buster; urgency=medium

  * Fix concurrent access due to missing locks in stats engine
    causing crashes.

 -- Atle Solbakken <atle@goliathdns.no>  Thu, 06 Aug 2020 00:15:03 +0200

rrr (1.7-3) buster; urgency=medium

  * Restart RRR service in post-installation script

 -- Atle Solbakken <atle@goliathdns.no>  Fri, 31 Jul 2020 00:05:51 +0200

rrr (1.7-2) buster; urgency=medium

  * Change default prefix of journal message topic by removing
    leading slash

 -- Atle Solbakken <atle@goliathdns.no>  Thu, 30 Jul 2020 22:33:28 +0200

rrr (1.7-1) buster; urgency=medium

  * Add ability to filter messages from senders by MQTT topic
    in all modules
  * Minor fixes

 -- Atle Solbakken <atle@goliathdns.no>  Wed, 29 Jul 2020 23:27:07 +0200

rrr (1.6-1) buster; urgency=medium

  * Add missing install file
  * Minor fixes

 -- Atle Solbakken <atle@goliathdns.no>  Mon, 27 Jul 2020 22:22:00 +0200


src/rrr_auth.c
/*

Read Route Record

Copyright (C) 2020 Atle Solbakken atle@goliathdns.no

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <termios.h>

#include "main.h"
#include "../build_timestamp.h"
#include "lib/rrr_config.h"
#include "lib/cmdlineparser/cmdline.h"
#include "lib/rrr_strerror.h"
#include "lib/version.h"
#include "lib/socket/rrr_socket.h"
#include "lib/log.h"
#include "lib/parse.h"
#include "lib/passwd.h"
#include "lib/util/linked_list.h"
#include "lib/util/gnu.h"
#include "lib/util/macro_utils.h"

RRR_CONFIG_DEFINE_DEFAULT_LOG_PREFIX("rrr_auth");

static const struct cmd_arg_rule cmd_rules[] = {
		{CMD_ARG_FLAG_NO_FLAG,		'\0',	"file",					"{PASSWORD_FILE}"},
		{CMD_ARG_FLAG_NO_FLAG,		'\0',	"username",				"{USERNAME}"},
		{CMD_ARG_FLAG_NO_ARGUMENT,	's',	"stdin",				"[-s|--stdin]"},
		{CMD_ARG_FLAG_HAS_ARGUMENT,	'p',	"permission",			"[-p|--permission[=]PERMISSION]"},
		{CMD_ARG_FLAG_NO_ARGUMENT,	'l',	"loglevel-translation",	"[-l|--loglevel-translation]"},
		{CMD_ARG_FLAG_HAS_ARGUMENT,	'e',	"environment-file",		"[-e|--environment-file[=]ENVIRONMENT FILE]"},
		{CMD_ARG_FLAG_HAS_ARGUMENT,	'd',	"debuglevel",			"[-d|--debuglevel[=]DEBUG FLAGS]"},
		{CMD_ARG_FLAG_HAS_ARGUMENT,	'D',	"debuglevel-on-exit",	"[-D|--debuglevel-on-exit[=]DEBUG FLAGS]"},
		{CMD_ARG_FLAG_NO_ARGUMENT,	'h',	"help",					"[-h|--help]"},
		{CMD_ARG_FLAG_NO_ARGUMENT,	'v',	"version",				"[-v|--version]"},
		{0,							'\0',	NULL,					NULL}
};

struct rrr_auth_data {
	int do_stdin;
	char *username;
	char *filename;
	char *permission;
};

static void __rrr_passwd_data_init (struct rrr_auth_data *data) {
	memset (data, '\0', sizeof(*data));
}

static void __rrr_passwd_destroy_data (struct rrr_auth_data *data) {
	RRR_FREE_IF_NOT_NULL(data->username);
	RRR_FREE_IF_NOT_NULL(data->filename);
	RRR_FREE_IF_NOT_NULL(data->permission);
}

static int __rrr_auth_parse_config (struct rrr_auth_data *data, struct cmd_data *cmd) {
	int ret = 0;

	const char *file = cmd_get_value(cmd, "file", 0);
	if (file != NULL && *file != '\0') {
		data->filename = strdup(file);
		if (data->filename == NULL) {
			RRR_MSG_0("Could not allocate memory in __rrr_auth_parse_config\n");
			ret = 1;
			goto out;
		}
	}

	const char *username = cmd_get_value(cmd, "username", 0);
	if (username != NULL && *username != '\0') {
		data->username = strdup(username);
		if (data->username == NULL) {
			RRR_MSG_0("Could not allocate memory in __rrr_auth_parse_config\n");
			ret = 1;
			goto out;
		}
	}

	if (data->filename == NULL || data->username == NULL || *(data->filename) == '\0' || *(data->username) == '\0') {
		RRR_MSG_0("Password filename and/or username not set\n");
		ret = 1;
		goto out;
	}

	const char *permission = cmd_get_value(cmd, "permission", 0);
	if (permission != NULL && *permission != '\0') {
		data->permission = strdup(permission);
		if (data->permission == NULL) {
			RRR_MSG_0("Could not allocate memory in __rrr_auth_parse_config\n");
			ret = 1;
			goto out;
		}
	}

	if (cmd_exists(cmd, "stdin", 0)) {
		data->do_stdin = 1;
	}
	else {
		data->do_stdin = 0;
	}

	out:
	return ret;
}

int main (int argc, const char **argv, const char **env) {
	if (!rrr_verify_library_build_timestamp(RRR_BUILD_TIMESTAMP)) {
		fprintf(stderr, "Library build version mismatch.\n");
		exit(EXIT_FAILURE);
	}

	// By default return failure
	int ret = EXIT_FAILURE;

	if (rrr_log_init() != 0) {
		goto out_final;
	}
	rrr_strerror_init();

	char *input_password = NULL;

	struct cmd_data cmd;
	struct rrr_auth_data data;

	cmd_init(&cmd, cmd_rules, argc, argv);
	__rrr_passwd_data_init(&data);

	if (rrr_main_parse_cmd_arguments_and_env(&cmd, env, CMD_CONFIG_DEFAULTS) != 0) {
		ret = EXIT_FAILURE;
		goto out;
	}

	// Don't require arguments here, separate check in parse_config
	if (rrr_main_print_help_and_version(&cmd, 0) != 0) {
		ret = EXIT_FAILURE;
		goto out;
	}

	if (__rrr_auth_parse_config(&data, &cmd) != 0) {
		cmd_print_usage(&cmd);
		ret = EXIT_FAILURE;
		goto out;
	}

	if (data.do_stdin) {
		if (rrr_passwd_read_password_from_stdin (&input_password) != 0) {
			ret = EXIT_FAILURE;
			goto out;
		}
	}
	else {
		if (rrr_passwd_read_password_from_terminal(&input_password, 0) != 0) {
			ret = EXIT_FAILURE;
			goto out;
		}
	}

	if (data.permission != NULL && *(data.permission) == '\0') {
		// Make sure empty permission is not \0 but NULL
		RRR_FREE_IF_NOT_NULL(data.permission);
	}

	if (rrr_passwd_authenticate(data.filename, data.username, input_password, data.permission) != 0) {
		RRR_MSG_0("Authentication failure\n");
		ret = EXIT_FAILURE;
		goto out;
	}
	else {
		RRR_DBG_1("Authentication succeded\n");
		ret = EXIT_SUCCESS;
	}

	goto out;

	out:
		rrr_config_set_debuglevel_on_exit();
		RRR_FREE_IF_NOT_NULL(input_password);
		__rrr_passwd_destroy_data(&data);
		cmd_destroy(&cmd);
		rrr_socket_close_all();
		rrr_strerror_cleanup();
		rrr_log_cleanup();
	out_final:
		return ret;
}
